File: frontend/src/features/auditoria/types/auditoria.ts
import { z } from "zod";

// ─── Acciones por Día ─────────────────────────────────────────────────────────

export const AccionPorDiaSchema = z.object({
  fecha: z.string(),
  total: z.number().int(),
}).passthrough();

export type AccionPorDia = z.infer<typeof AccionPorDiaSchema>;

// ─── Comunicaciones por Docente ───────────────────────────────────────────────

export const ComunicacionPorDocenteSchema = z.object({
  usuario_id: z.string().uuid(),
  nombre: z.string(),
  Pendiente: z.number().int().optional(),
  Enviando: z.number().int().optional(),
  OK: z.number().int().optional(),
  Fallido: z.number().int().optional(),
  Cancelado: z.number().int().optional(),
}).passthrough();

export type ComunicacionPorDocente = z.infer<typeof ComunicacionPorDocenteSchema>;

// ─── Interacciones por Docente-Materia ────────────────────────────────────────

export const InteraccionPorDocenteMateriaSchema = z.object({
  usuario_id: z.string().uuid(),
  nombre: z.string(),
  materia_id: z.string().uuid(),
  materia_nombre: z.string(),
  acciones: z.record(z.string(), z.number().int()),
  total: z.number().int(),
}).passthrough();

export type InteraccionPorDocenteMateria = z.infer<
  typeof InteraccionPorDocenteMateriaSchema
>;

// ─── Última Acción ────────────────────────────────────────────────────────────

export const UltimaAccionSchema = z.object({
  id: z.string().uuid(),
  fecha_hora: z.string().nullable().optional(),
  actor_nombre: z.string().nullable().optional(),
  accion: z.string(),
  materia_nombre: z.string().nullable().optional(),
  detalle: z.any().optional(),
  ip: z.string().nullable().optional(),
}).passthrough();

export type UltimaAccion = z.infer<typeof UltimaAccionSchema>;

// ─── Log Item ─────────────────────────────────────────────────────────────────

export const LogItemSchema = z.object({
  id: z.string().uuid(),
  fecha_hora: z.string().nullable().optional(),
  actor_id: z.string().uuid().nullable().optional(),
  actor_nombre: z.string().nullable().optional(),
  materia_id: z.string().uuid().nullable().optional(),
  materia_nombre: z.string().nullable().optional(),
  accion: z.string(),
  detalle: z.any().optional(),
  filas_afectadas: z.number().int().nullable().optional(),
  ip: z.string().nullable().optional(),
  user_agent: z.string().nullable().optional(),
}).passthrough();

export type LogItem = z.infer<typeof LogItemSchema>;

export const LogResponseSchema = z.object({
  items: z.array(LogItemSchema),
  total: z.number().int(),
}).passthrough();

export type LogResponse = z.infer<typeof LogResponseSchema>;

// ─── Filtros ──────────────────────────────────────────────────────────────────

export interface AuditoriaFilters {
  fecha_desde?: string;
  fecha_hasta?: string;
  materia_id?: string;
}

export interface LogFilters extends AuditoriaFilters {
  usuario_id?: string;
  accion?: string;
  offset?: number;
  limit?: number;
}

File: backend/app/schemas/auditoria.py
"""Schemas Pydantic para el panel de auditoría y métricas (C-19).

Todos los schemas usan ``extra='forbid'`` (REGLAS DURAS #5).
"""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Panel de interacciones (F9.1) ──────────────────────────────────


class AccionesPorDiaItem(BaseModel):
    """Agregación de acciones por día."""

    model_config = ConfigDict(extra="forbid")

    fecha: date
    total: int


class ComunicacionesPorDocenteItem(BaseModel):
    """Distribución de estados de comunicación por docente."""

    model_config = ConfigDict(extra="forbid")

    usuario_id: UUID
    nombre: str
    Pendiente: int = 0
    Enviando: int = 0
    OK: int = 0
    Fallido: int = 0
    Cancelado: int = 0


class InteraccionesItem(BaseModel):
    """Agregación de interacciones por docente y materia."""

    model_config = ConfigDict(extra="forbid")

    usuario_id: UUID
    nombre: str
    materia_id: UUID
    materia_nombre: str
    acciones: dict[str, int]
    total: int


class UltimasAccionesItem(BaseModel):
    """Últimas acciones registradas en el sistema."""

    model_config = ConfigDict(extra="forbid")

    id: UUID
    fecha_hora: datetime
    actor_nombre: str
    accion: str
    materia_nombre: str | None = None
    detalle: dict | None = None
    ip: str | None = None


# ── Log completo de auditoría (F9.2) ───────────────────────────────


class LogItem(BaseModel):
    """Registro individual del log completo de auditoría."""

    model_config = ConfigDict(extra="forbid")

    id: UUID
    fecha_hora: datetime
    actor_id: UUID
    actor_nombre: str
    materia_id: UUID | None = None
    materia_nombre: str | None = None
    accion: str
    detalle: dict | None = None
    filas_afectadas: int | None = None
    ip: str | None = None
    user_agent: str | None = None


class LogPaginado(BaseModel):
    """Respuesta paginada del log de auditoría."""

    model_config = ConfigDict(extra="forbid")

    items: list[LogItem]
    total: int
    offset: int
    limit: int


# ── Filtros ────────────────────────────────────────────────────────


class FiltrosAuditoria(BaseModel):
    """Filtros combinables para consultas de auditoría.

    Todos los campos son opcionales — solo se aplican los presentes.
    """

    model_config = ConfigDict(extra="forbid")

    fecha_desde: date | None = None
    fecha_hasta: date | None = None
    materia_id: UUID | None = None
    usuario_id: UUID | None = None
    accion: str | None = None
